src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    stateHistory.push(currentState);
```

Comment:
To avoid potential issues with object references, you should push a copy of the `currentState` object to the `stateHistory` array instead of the `currentState` object itself. This ensures that each element in the `stateHistory` array is a distinct object. You can create a copy using the spread operator `{ ...currentState }`.